# AdvCiv-SAS reachable-history commit diff
# Email privacy: email-shaped addresses redacted from generated commit-history text.
# History metadata layout: 3 (history-segment label; title-only diff header; full message redirected to anonymized Git logs).
# Cache format: 2
# Cache policy: 8c87d6e4bd0f
# Practical commit count: 3239
# Commit: 4e6fe9df58dfbce36a575aa24224e44030bd76fa
# History segment: pre-SAS AdvCiv history
# Parent used for diff: e397ad6890a8e5d65a854da0019603203442b1a9
# Title: Redistribute rounding errors from non-binary research
# Full commit message/metadata: see _0_Common_Docs/git_logs/git_log_anonymized_email_002_Base_AdvCiv.txt.
# History segments describe ancestry/log partition, not authorship: SASBranch includes SAS work plus later upstream AdvCiv commits merged/imported into this branch.
# Practical commit counts can repeat on divergent/merged history; the full Git SHA is the canonical unique commit identifier.
# Diff policy: preserve textual source/docs/config history; summarize known generated/binary/noisy or exceptionally huge file patches.
# Git textconv/external diff drivers are disabled; redundant document formats such as ODT/PDF are summarized rather than converted.

[CHANGE SUMMARY - NUMSTAT + RENAME/MODE SUMMARY]
3	2	CvGameCoreDLL/CvCity.cpp
65	22	CvGameCoreDLL/CvPlayer.cpp
10	3	CvGameCoreDLL/CvPlayer.h

[PATCHES]

diff --git a/CvGameCoreDLL/CvCity.cpp b/CvGameCoreDLL/CvCity.cpp
index 1fb32660e..c3b11b6bb 100644
--- a/CvGameCoreDLL/CvCity.cpp
+++ b/CvGameCoreDLL/CvCity.cpp
@@ -4696,7 +4696,7 @@ void CvCity::changeGovernmentCenterCount(int iChange)
 	Doesn't check if the building can be constructed in this city. */
 int CvCity::getSavedMaintenanceTimes100ByBuilding(BuildingTypes eBuilding) const
 {
-	CvBuildingInfo& kBuilding = GC.getInfo(eBuilding);
+	CvBuildingInfo const& kBuilding = GC.getInfo(eBuilding);
 	int iModifier = kBuilding.getMaintenanceModifier();
 	if (iModifier != 0 && !isNoMaintenance())
 	{
@@ -7200,7 +7200,8 @@ void CvCity::updateCommerce(CommerceTypes eCommerce)
 
 		m_aiCommerceRate.set(eCommerce, iNewRate);
 		FAssert(m_aiCommerceRate.get(eCommerce) >= 0);
-		GET_PLAYER(getOwner()).changeCommerceRate(eCommerce, iNewRate - iOldRate);
+		GET_PLAYER(getOwner()).changeCommerceRateTimes100(eCommerce,
+				iNewRate - iOldRate);
 	}
 	GET_PLAYER(getOwner()).invalidateCommerceRankCache(eCommerce);
 	if (isCitySelected())

diff --git a/CvGameCoreDLL/CvPlayer.cpp b/CvGameCoreDLL/CvPlayer.cpp
index e0b060bae..c8868052e 100644
--- a/CvGameCoreDLL/CvPlayer.cpp
+++ b/CvGameCoreDLL/CvPlayer.cpp
@@ -457,7 +457,8 @@ void CvPlayer::reset(PlayerTypes eID, bool bConstructorCall)
 	m_aiTradeYieldModifier.reset();
 	m_aiFreeCityCommerce.reset();
 	m_aiCommercePercent.reset();
-	m_aiCommerceRate.reset();
+	m_aiCommerceRateTimes100.reset();
+	m_aiCommerceRate.reset(); // advc.157
 	m_aiCommerceRateModifier.reset();
 	m_aiCapitalCommerceRateModifier.reset();
 	m_aiStateReligionBuildingCommerce.reset();
@@ -9666,33 +9667,69 @@ bool CvPlayer::setCommercePercent(CommerceTypes eCommerce, int iNewValue, bool b
 	return true;
 }
 
-
-int CvPlayer::getCommerceRate(CommerceTypes eCommerce) const
+/*	advc.157: Based on the BtS getCommerceRate(CommerceTypes) function.
+	The part that redistributes rounding errors is new - and the reason
+	why all all commerce rates are updated at once. */
+void CvPlayer::updateCommerceRates()
 {
-	int iRate = m_aiCommerceRate.get(eCommerce);
-	if (GC.getGame().isOption(GAMEOPTION_NO_ESPIONAGE))
+	/*	advc.test: In CvPlayer::updateCommerce, it should suffice to call
+		updateCommerceRates only once in the end and to suppress the calls
+		currently coming from CvCity. Well, let's first see if there's any
+		performance problem. */
+	PROFILE_FUNC();
+
+	EagerEnumMap<CommerceTypes,int,short> aiError;
+	int iTotalError = 0;
+	FOR_EACH_ENUM2(Commerce, eCommerce)
 	{
-		if (eCommerce == COMMERCE_CULTURE)
-			iRate += m_aiCommerceRate.get(COMMERCE_ESPIONAGE);
-		else if (eCommerce == COMMERCE_ESPIONAGE)
-			iRate = 0;
-	}  // <advc.004x>
-	iRate /= 100;
-	if (eCommerce == COMMERCE_RESEARCH)
-	{	// advc.910:
-		static int const iBASE_RESEARCH_RATE = GC.getDefineINT("BASE_RESEARCH_RATE");
-		iRate += iBASE_RESEARCH_RATE;
+		int iRate = m_aiCommerceRateTimes100.get(eCommerce);
+		if (GC.getGame().isOption(GAMEOPTION_NO_ESPIONAGE))
+		{
+			if (eCommerce == COMMERCE_CULTURE)
+				iRate += m_aiCommerceRateTimes100.get(COMMERCE_ESPIONAGE);
+			else if (eCommerce == COMMERCE_ESPIONAGE)
+				iRate = 0;
+		}
+		{
+			int iError = (iRate < 0 ? -1 : 1) * (abs(iRate) % 100);
+			aiError.set(eCommerce, iError);
+			iTotalError += iError;
+		}
+		iRate /= 100;
+		m_aiCommerceRate.set(eCommerce, iRate);
 	}
-	return iRate; // </advc.004x>
+	// Implicit selection sort (b/c the error is going to be small)
+	for (int i = 0; i < abs(iTotalError) / 100; i++)
+	{
+		/*	Not sure if negative error needs to be supported here;
+			maintenance isn't included. Well, let's future-proof. */
+		int const iSign = (iTotalError > 0 ? 1 : -1);
+		CommerceTypes eArgMax = NO_COMMERCE;
+		int iMaxAbsError = MIN_INT;
+		FOR_EACH_ENUM2(Commerce, eCommerce)
+		{
+			int const iError = aiError.get(eCommerce) * iSign;
+			if (iError > iMaxAbsError)
+			{
+				eArgMax = eCommerce;
+				iMaxAbsError = iError;
+			}
+		}
+		m_aiCommerceRate.add(eArgMax, iSign * 1);
+	}
+	// <advc.004x>
+	static int const iBASE_RESEARCH_RATE = GC.getDefineINT("BASE_RESEARCH_RATE"); // advc.910
+	m_aiCommerceRate.add(COMMERCE_RESEARCH, iBASE_RESEARCH_RATE); // </advc.004x>
 }
 
 
-void CvPlayer::changeCommerceRate(CommerceTypes eCommerce, int iChange)
+void CvPlayer::changeCommerceRateTimes100(CommerceTypes eCommerce, int iChange)
 {
 	if (iChange != 0)
 	{
-		m_aiCommerceRate.add(eCommerce, iChange);
+		m_aiCommerceRateTimes100.add(eCommerce, iChange);
 		FAssert(getCommerceRate(eCommerce) >= 0);
+		updateCommerceRates(); // advc.157
 		if (isActive())
 			gDLL->UI().setDirty(GameData_DIRTY_BIT, true);
 	}
@@ -13997,7 +14034,11 @@ void CvPlayer::read(FDataStreamBase* pStream)
 		m_aiTradeYieldModifier.read(pStream);
 		m_aiFreeCityCommerce.read(pStream);
 		m_aiCommercePercent.read(pStream);
-		m_aiCommerceRate.read(pStream);
+		m_aiCommerceRateTimes100.read(pStream);
+		// <advc.157>
+		if (uiFlag >= 17)
+			m_aiCommerceRate.read(pStream);
+		else updateCommerceRates(); // </advc.157>
 		m_aiCommerceRateModifier.read(pStream);
 		m_aiCapitalCommerceRateModifier.read(pStream);
 		m_aiStateReligionBuildingCommerce.read(pStream);
@@ -14013,7 +14054,7 @@ void CvPlayer::read(FDataStreamBase* pStream)
 		m_aiTradeYieldModifier.readArray<int>(pStream);
 		m_aiFreeCityCommerce.readArray<int>(pStream);
 		m_aiCommercePercent.readArray<int>(pStream);
-		m_aiCommerceRate.readArray<int>(pStream);
+		m_aiCommerceRateTimes100.readArray<int>(pStream);
 		m_aiCommerceRateModifier.readArray<int>(pStream);
 		m_aiCapitalCommerceRateModifier.readArray<int>(pStream);
 		m_aiStateReligionBuildingCommerce.readArray<int>(pStream);
@@ -14382,7 +14423,8 @@ void CvPlayer::write(FDataStreamBase* pStream)
 	//uiFlag = 13; // advc.004s
 	//uiFlag = 14; // advc.027 (m_bRandomWBStart)
 	//uiFlag = 15; // advc.908a (separate tag for nerfed trait effect), advc.908c
-	uiFlag = 16; // advc.enum: new enum map save behavior
+	//uiFlag = 16; // advc.enum: new enum map save behavior
+	uiFlag = 17; // advc.157
 	pStream->Write(uiFlag);
 
 	// <advc.027>
@@ -14521,7 +14563,8 @@ void CvPlayer::write(FDataStreamBase* pStream)
 	m_aiTradeYieldModifier.write(pStream);
 	m_aiFreeCityCommerce.write(pStream);
 	m_aiCommercePercent.write(pStream);
-	m_aiCommerceRate.write(pStream);
+	m_aiCommerceRateTimes100.write(pStream);
+	m_aiCommerceRate.write(pStream); // advc.157
 	m_aiCommerceRateModifier.write(pStream);
 	m_aiCapitalCommerceRateModifier.write(pStream);
 	m_aiStateReligionBuildingCommerce.write(pStream);

diff --git a/CvGameCoreDLL/CvPlayer.h b/CvGameCoreDLL/CvPlayer.h
index 376ff1c71..d8629b11b 100644
--- a/CvGameCoreDLL/CvPlayer.h
+++ b/CvGameCoreDLL/CvPlayer.h
@@ -937,8 +937,13 @@ public:
 		return setCommercePercent(eCommerce, getCommercePercent(eCommerce) + iChange);
 	}
 
-	int getCommerceRate(CommerceTypes eCommerce) const;																// Exposed to Python
-	void changeCommerceRate(CommerceTypes eCommerce, int iChange);
+	int getCommerceRate(CommerceTypes eCommerce) const																// Exposed to Python
+	{
+		return m_aiCommerceRate.get(eCommerce); // advc.157: cached
+	}
+	void updateCommerceRates(); // advc.157
+	// advc.157: Renamed from "changeCommerceRate"
+	void changeCommerceRateTimes100(CommerceTypes eCommerce, int iChange);
 
 	int getCommerceRateModifier(CommerceTypes eCommerce) const { return m_aiCommerceRateModifier.get(eCommerce); }	// Exposed to Python
 	void changeCommerceRateModifier(CommerceTypes eCommerce, int iChange);
@@ -1588,7 +1593,9 @@ protected:  // <advc.210>
 	YieldPercentMap m_aiTradeYieldModifier;
 	CommerceChangeMap m_aiFreeCityCommerce;
 	CommercePercentMap m_aiCommercePercent;
-	EagerEnumMap<CommerceTypes,int> m_aiCommerceRate; // (at times-100 precision)
+	EagerEnumMap<CommerceTypes,int> m_aiCommerceRateTimes100;
+	// advc.157: Cache actual rate (not times 100) separately
+	EagerEnumMap<CommerceTypes,int,short> m_aiCommerceRate;
 	CommercePercentMap m_aiCommerceRateModifier;
 	CommercePercentMap m_aiCapitalCommerceRateModifier;
 	CommerceChangeMap m_aiStateReligionBuildingCommerce;

[PATCH COVERAGE]
patchFiles=3 embeddedFiles=3 omittedFiles=0 embeddedPatchBytes=8077
